From ab4896899c8a1137dcc97d5fa73aaf6faeaf3674 Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Mon, 14 Jul 2014 13:41:37 -0400
Subject: [x86] mce: Call put_device on device_register failure

Message-id: <1405345303-24468-8-git-send-email-prarit@redhat.com>
Patchwork-id: 85778
O-Subject: [RHEL7.1 PATCH BZ 1119306 07/13] x86, mce: Call put_device on device_register failure
Bugzilla: 1119306
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1119306

commit 853d9b18f1e861d37e9b271742329f8c1176eabe
Author: Levente Kurusa <levex@linux.com>
Date:   Fri Nov 29 21:28:48 2013 +0100

    x86, mce: Call put_device on device_register failure

    This patch adds a call to put_device() when the device_register() call
    has failed. This is required so that the last reference to the device is
    given up.

    Signed-off-by: Levente Kurusa <levex@linux.com>
    Link: http://lkml.kernel.org/r/5298F900.9000208@linux.com
    Signed-off-by: Borislav Petkov <bp@suse.de>

Cc: Tony Camuso <tcamuso@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kernel/cpu/mcheck/mce.c b/arch/x86/kernel/cpu/mcheck/mce.c
index 10f609a..9b7734b 100644
--- a/arch/x86/kernel/cpu/mcheck/mce.c
+++ b/arch/x86/kernel/cpu/mcheck/mce.c
@@ -2272,8 +2272,10 @@ static int mce_device_create(unsigned int cpu)
  dev->release = &mce_device_release;
 
  err = device_register(dev);
- if (err)
+ if (err) {
+  put_device(dev);
   return err;
+ }
 
  for (i = 0; mce_device_attrs[i]; i++) {
   err = device_create_file(dev, mce_device_attrs[i]);
-- 
1.7.1